### ISSUE TITLE
Bug: no-misused-promises Optional chaining void rule

### ISSUE THREAD

**Issue by nikolai-tise** [typescript-eslint/typescript-eslint#7637]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=DYUwLgBAZg9jBcEDeAjAhgJ0QCgJQQF4A%2BCABQxgFsBLAZxAB40A7ATyIF8IAfCAV2YATEFGrMQggNwAoalAjYAsmjAALAHQYWgqnggkADOoCM%2BJNIjQ4hZBcsQ0tVswDGEdBjyJyVOoxbsyBx2wcFyCrAwAPzqHnhmEAD0iRBqdBC0qjB8wIIQqmgAbiCpMO4laADuaNRgEtIcQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUFs%2B5AAgC4CeADigMbQCWphAtCvNQHaED0LA9vQLbXJYyRABN6paF37DUGSImiTokcGAC%2BINUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
let foo: {bar: () => Promise<any>} | undefined;
if (Math.random() > 0.1) {
  foo = {
    async bar(): Promise<any> {}
  }
}
if (foo?.bar()) { // this should have to be awaited
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
     "@typescript-eslint/no-misused-promises": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

I expected that the `foo?.bar()` call should have to be awaited

### Actual Result

There was no error forcing this to be awaited

**Comment by nikolai-tise** [comment]
Same goes for the `no-floating-promises` rule

**Comment by JoshuaKGoldberg** [comment]
A bit simpler of a repro that shows the rule not complaining in either case of the function being nullish _or_ the function returning `Promise<unknown> | undefined`:

```ts
declare const nullishAsyncFunction: (() => Promise<unknown>) | undefined;

if (nullishAsyncFunction?.()) { }

declare const functionReturnsNullishAsync: () => (Promise<unknown> | undefined);

if (functionReturnsNullishAsync()) { }
```

I could see a use case for wanting to check whether a function produced a Promise or not... but it feels like that's a bug (so, agreeing with the issue 😄). Maybe we should add an option to the rule? We could make it opt-in by default for now, and opt-out in the next major version?   

> `no-floating-promises`

Would you be open to filing a separate issue? I think we'd want to track them separately.

For both, cc @typescript-eslint/triage-team - would like to hear more opinions.

**Comment by nikolai-tise** [comment]
> I could see a use case for wanting to check whether a function produced a Promise or not... but it feels like that's a bug 
> (so, agreeing with the issue 😄). Maybe we should add an option to the rule? We could make it opt-in by default for now, 
> and opt-out in the next major version?

That would be awesome, thanks! 😄

> no-floating-promises

> Would you be open to filing a separate issue? I think we'd want to track them separately.

Sure no problem!

**Comment by bradzacher** [comment]
I don't remember but I vaguely remember this being intentional design?
It's a really complex case because there are arguments in both directions for this.

One could say "oh but optional chaining" - but what's the difference between

```ts
function foo(): Promise | undefined
if (foo()) {} // is this misused?

const bar: {baz: Promise} | undefined
if (bar?.baz) {} // what about this? It has the same type!
```

Should one be valid but not the other? Seems weird if it is cos the types are the same - would the intent not also be the same?

What about this:

```ts
const maybeProm: Promise | false;
if (maybeProm) {}
```

Is this a misused promise? The guard does narrow things and have meaning!

Depending on your perspective either could be correct!

**Comment by JoshuaKGoldberg** [comment]
Coming back to this: I think it'd be reasonable to add an opt-in option to the rule. I can see the use case for folks wanting to check whether a nullable Promise is truthy or not. If they're using this rule with [`@typescript-eslint/strict-boolean-expression`](https://typescript-eslint.io/rules/strict-boolean-expressions) then that should capture most of the issues. Thoughts?

**Comment by Josh-Cena** [comment]
I'm extremely split on this. On one hand I can imagine some use cases for this, particularly such as a class that "maybe contains a pending state, maybe not". On the other hand we can't assume idempotency between function calls, so even if the first call produces a promise, there's nothing we can say about the second, and this seems like a developer mistake as they really should have assigned the promise to a variable first (also async functions usually have side-effects so calling an async function twice is even worse). All this said, I think a call expression producing a promise should be an error by default because usually the promise is produced fresh each time.

**Comment by kirkwaiblinger** [comment]
I'd be +1 to considering a promise used in a boolean context to be misused by default.
However, a la https://github.com/typescript-eslint/typescript-eslint/issues/7637#issuecomment-1721967558, it may not be possible to introduce this check without introducing some false positives. Maybe those are niche enough that they can just be deal with using eslint-disable-next-line comments?

**Comment by Josh-Cena** [comment]
Assuming the type is `if (Promise<boolean> | undefined) { do something }`:

- `if (maybePromise)` - this is (almost) definitely intentional
- `if (obj.maybePromise)` - I would say it's intentional
- `if (maybeObj?.definitelyPromise)` - I would say it's intentional
- `if (f())`, `if (obj.f())`, `if (obj?.f())`, `if (f?.())` - I would say it's a mistake

Does that make sense? If I'm to design the logic, I would say that if we find a `Promise | maybeFalsy` value used in a boolean context, we should only report if this value is produced from a call expression.

**Comment by Josh-Cena** [comment]
Actually this is a duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/5767. I think an option would do
